google/cloud/storage/internal/generic_stub_factory.cc

```diff
@@ -0,0 +1,63 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/generic_stub_factory.h"
+#include "google/cloud/storage/internal/curl/stub.h"
+#include "google/cloud/storage/internal/logging_stub.h"
+#include "google/cloud/storage/internal/rest/stub.h"
+#include "google/cloud/storage/options.h"
+#include "google/cloud/internal/algorithm.h"
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+bool RequiresLogging(Options const& opts) {
+  using ::google::cloud::internal::Contains;
+  auto const& tracing_components = opts.get<TracingComponentsOption>();
+  return Contains(tracing_components, "raw-client") ||
+         Contains(tracing_components, "rpc");
+}
+
+std::unique_ptr<GenericStub> DecorateStub(bool logging,
+                                          std::unique_ptr<GenericStub> stub) {
+  if (logging) {
+    stub = std::make_unique<storage::internal::LoggingStub>(std::move(stub));
+  }
+  return stub;
+}
+
+}  // namespace
+
+std::unique_ptr<GenericStub> DecorateStub(Options const& opts,
+                                          std::unique_ptr<GenericStub> stub) {
+  return DecorateStub(RequiresLogging(opts), std::move(stub));
+}
+
+std::unique_ptr<GenericStub> MakeDefaultStorageStub(Options opts) {
```

Comment:
This function is being fed `Options const&`: https://github.com/googleapis/google-cloud-cpp/blob/441c3b620b4a2c046cd006141f2968cbbcd7d806/google/cloud/storage/internal/connection_factory.cc#L44-L47

I think it would simplify the code a smidge if we accept `Options const&` here. WDYT?

```cc
std::unique_ptr<GenericStub> MakeDefaultStorageStub(Options const& opts) {
  auto stub = [&opts] -> std::unique_ptr<GenericStub> {
    if (opts.get<storage::internal::UseRestClientOption>()) {
      return std::make_unique<storage::internal::RestStub>(opts);
    }
    return std::make_unique<storage::internal::RestStub>(opts);    
  }();
  return DecorateStub(RequiresLogging(opts), std::move(stub));
}
```

Reply:
Not sure: (1) the root factory function for `*Stub` in other libraries consumes a `Options opts`. (2) the copy of the options has to happen somewhere.  As-is, it happens at the `MakeDefaultStorageStub()` call site. As proposed it would happen as part of the `std::make_unique<>` calls. (3) I can read the code in the PR and in your changes just fine.

I am leaving it as-is. easy to change if we think it was the wrong call.